### PR TITLE
At module

### PR DESCRIPTION
**Pull request by risaacson**
I have a usage story that requires the delayed execution of a command or script at some point in the future relating to testing or notifications. The 'at' command is a standard Unix tool that is often overlooked in relation to its usefulness.

### CODE REVIEW

File: library/system/at [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        fileh.write(command)
+        fileh.close()
+        at_command = "%s now + %s %s -f %s" % (ATCMD, unit_count, unit_type, path)
+        if user:
+            at_command = "chown %s %s; su '%s' -c '%s'" % (user, path, user, at_command)
+        rc, out, err = module.run_command(at_command)
+        if rc != 0:
+            module.fail_json(msg=err)
+        os.unlink(path)
+        result['changed'] = 'true'
```
Comment (resmo) [inline library/system/at]: this should be boolean `result['changed'] = True`

File: library/system/at [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            unit_count=dict(required=True),
+            unit_type=dict(required=True,
+            	           default=None,
+            	           choices=["minutes", "hours", "days", "weeks"],
+            	           type="str")
+        ),
+        supports_check_mode = False,
+    )
+
+    user        = module.params['user']
```
Comment (resmo) [inline library/system/at]: minor indention fix?

File: library/system/at [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    elif script_file:
+        result['script_file'] = script_file
+        at_command = "%s now + %s %s -f %s" % (ATCMD, unit_count, unit_type, script_file)
+        if user:
+        	# We expect that if this is an installed the permissions are already correct for the user to execute it.
+            at_command = "su '%s' -c '%s'" % (user, at_command)
+        rc, out, err = module.run_command(at_command)
+        if rc != 0:
+            module.fail_json(msg=err)
+        result['changed'] = 'true'
```
Comment (resmo) [inline library/system/at]: same here as for line 115

**Comment by risaacson** [comment]
Thanks for catching those @resmo. I have pushed the fixes and scolded my editor for sneaking in a tabs where spaces should be.

**Comment by jctanner** [comment]
@risaacson rather than setting a global for the bin path, try using module.get_bin_path().

**Comment by resmo** [comment]
to me it seems if as the module would always create new at jobs. But the whole point of ansible is, that it detects the state and handles accordingly: 
- no such job, create it --> result['changed'] = True, 
- at job exists, nothing to do here --> result[''changed] = False

would you implement functionality?

**Comment by mpdehaan** [comment]
Hi Richard,

Thanks for this -- can you explain the logic for why it creates a tempfile to support the command option?

It seems like it could just schedule the command directly, possibly?

As for the above, I am not entirely sure neccessarily that this module needs to implement changed True/False correctly initially as it might be somewhat tricky to parse the list of at jobs.

**Comment by risaacson** [comment]
@jctanner thanks for pointing that out. I had just seen that used in another one of the modules but it had not been back-ported to the module that I was using as a rough template. I'll upgrade that part of the code and bring it in line.

@resmo thank you also for your comment. A similar question was in my head as I was putting together the module as I was trying to work out what a wider audience would want from the module. From the common usage that I have seen at tends to be a "set and forget" task where uniqueness or the ability to recall(delete) is rarely used. In addition the at command has a short lifetime having state. The entry lives in the queue for a short period and then is executed removing which destroys the state. This is in contrast to a crontab entry or a cron.d file which has state until you explicitly remove it. Keeping that in mind I delivered a basic version of the plugin with what I have evaluated should be the default behavior from common usage.

@resmo with the thinking above stated you have jumped ahead to what I think should be the v0.2 version of the plugin which would give the consumer more control. To combine our stories it seems that the evolution of the module would be to add a "state" option consisting of either create(default), unique, and delete. The create functionality is already implemented. The unique functionality covers exactly what you are saying to make sure that if there is an existing entry as the user that it is not queued again. Since you have to build in the ties to "atq" and "at -c JOBNUM" the delete functionality becomes just about free.

@resmo TL:DR? ;) at is a blackbox service rather then a daemon that you configure. The functionality could be useful though to create unique jobs or remove existing jobs. Does this seem in line with what you are looking for? If so I will implement it this evening.

@mpdehaan easy answer for that question. The reason to use tempfile is that it makes only one path to creating a job and it will be easier to refactor down and shrink the code. I am also lazy because to add a job that is not in a  file requires you to use at interactively as there is not a space on the acceptable commands for a string representing the command. It just isn't implemented in the versions of at that I have tested on.

**Comment by risaacson** [comment]
I went ahead and worked on the implementation as described as there are some cases that it would be useful. Testing is coming back clean.

**Comment by mpdehaan** [comment]
@risaacson makes sense if you can't pass arguments.   Looking at testing this now.

**Comment by mpdehaan** [comment]
Ok review notes

Rather than using custom code to check for mutual exclusivity, please use "mutually_exclusive" and "required_one_of" as available in module_common support code.  You should be able to find this listed in a few other modules.

The "action" should be more mapped to "state", and take values like "present" or "absent" 

"unique" should be a boolean parameter, only valid when set with present, that has a 'type='bool''

The type='str' should be specified for other module parameters.

The 'action' variable should not appear in the return results.   It's ok if 'state' does.

Invoking "su" seems like it might prompt for a password.   Su support shouldn't be done here, but the user should su or sudo using Ansible, so remove su support.

It appears the "action==unique" codeblock might need some error handling, as I see where it calls "exit_json" but should probably use some function do decide if it should call fail_json instead.  While it is true that including "failed=True" in a result is valid, we want to do this to make the code easier to read.

I'd probably rename "counts" to "count" and "unit_type" to "units", so you could closer to English say "count=2 units=minutes" vs "units=2 unit_type=minutes"

module.run_command has flags so failure on rc !=0 doesn't need to be explicitly handled.

Update the "version_added" field to say "1.5" vs "0.0", also a few minor typos.

The unit type should probably have a default.  Minutes seems reasonable.

Rather than have a long-form main function, move the implementation of things like delete into smaller functions.

Otherwise, looks good, so if you can update this, let me know, and we can test it out.

This appears to function pretty much like we want, so it's just a matter of getting the code polished up to use all the built in magic, and in particular remove the 'su' stuff as that could be problematic.

Ansible will likely natively grok 'su' instead of just 'sudo' in the near future (pending pull request, but some tweaks might need to be applied too).

**Comment by risaacson** [comment]
I did end up using mutually_exclusive and required_one_of for command and script_file. Also I switched unique to a Boolean and added a check so that it was not used with state absent.

I am going to remove the user option as I was wavering about it though it is in for the cron module. I agree that there are better ways to accomplish the same goal.

The action==unique block is really just a modifier and I don't think that it can fail. Either it exits because the command is not unique or it lets the rest of the script run. Child failures are caught in the match_jobs method. Though by bringing this up it points out that it would leave a script file sitting on the filesystem rather then cleaning it up.

I renamed the unit_count to count and unit_type to units. Update the run_command to use the check_rc.

I had not put a default for the unit type as it fits with the at command that it is specified. I like the idea of enforcing clarity here.

Commits pushed!

**Comment by mpdehaan** [comment]
Looks good to me, in queue for testing, thanks for updates!

**Comment by jimi-c** [comment]
Closing, as this appears to be an old version of the PR. The at module has been merged in via a later PR.

**Pull request closed without merging**